[1기-P] 김부희 - JPA 게시판 구현하기

buri-1029 opened the pull request:
# Spring Boot JPA로 게시판 구현
> Spring Boot JPA - Rest API를 강의를 듣고, 게시판 구현 미션 수행
> [리드미](https://github.com/buri-1029/SpringBoot-Board/blob/main/README.md)

## Entity 연관관계 설정 ✔︎ 
   - User(사용자) ⬅️ Post(게시글) `1:N`, `단방향`
   - Post(게시글) ↔️ Comment(댓글) `1:N`, `양방향`

##  API 구현 ✔︎ 

## REST Docs 이용한 문서화 ✔︎ 

---

## 구현내용(요구사항)
  - 게시글과 댓글은 해당 작성자만 수정하고 삭제할 수 있습니다. 
  - 사용자가 탈퇴해도 게시글은 남아있습니다.
  - 게시글 삭제 시 댓글까지 함께 삭제됩니다.
  - userId가 담긴 dto를 통해 유저정보를 전달하여 가입 유저를 확인하는 방식으로 구현하였습니다. (+ API를 Restful 하도록 구현하고 싶기도 하여서 userId를 파라미터로 넘기지 않았음)

buri-1029 committed:
base setting

buri-1029 committed:
feat : 엔티티, 연관관계 매핑 설계

buri-1029 committed:
test : Repository 테스트 작성

buri-1029 committed:
refactor : 생성자 파라미터 변경

buri-1029 committed:
feat : User, Post, Comment Service 단 작성

buri-1029 committed:
test : Service 테스트 작성

buri-1029 committed:
refactor : 코드 리팩토링

불필요한 코드 제거,, requestDto에 기본 생성자 추가

buri-1029 committed:
refactor : Post, Commit Entity 필드 변경

buri-1029 committed:
feat : User, Post, Comment Contoller 단 작성

buri-1029 committed:
feat : rest docs 문서화

buri-1029 committed:
style : continuation indent 4로 변경

buri-1029 committed:
refactor : 코드 리팩토링

패키지구조 변경, response 구조 커스텀

buri-1029 committed:
feat : dto 필드 not null 검증 추가

buri-1029 committed:
refactor : PostController API 수정, README.md 수정

buri-1029 committed:
.DS_Store removed

neilsonT commented on springboot-board/src/main/java/org/prgms/board/comment/controller/CommentController.java:
```diff
@@ -0,0 +1,34 @@
+        this.commentService = commentService;
+    }
+
+    @PostMapping("/{userId}/{postId}")
+    public ApiResponse<Long> addComment(@PathVariable Long userId, @PathVariable Long postId, @RequestBody @Valid CommentRequest commentRequest) {
+        return ApiResponse.toResponse(commentService.addComment(userId, postId, commentRequest));
+    }
```
URI가 리소스를 표현한다고 알고 있는데, /api/comments/1 이라면, 1이 댓글 번호라고 생각할 수도 있을 것 같네요.
이 부분은 저도 많이 헷갈려서 의견을 듣고 싶습니다 ㅎ.
저라면, /api/comments?userId=2&postId=3 이런식으로 작성하고 @RequestParam을 사용했을 것 같아요.

buri-1029 replied:
넹 저도 이 부분에 대해서 둘 중 무엇을 사용할지 고민을 했는데요.  Restful 서비스에서 사용되는 URL의 형태는 /api/{id} 이런 형식이라 무작정 `@PathVariable` 를 사용했는데 더 찾아보니 `@PathVariable`은  {userId}와 {postId}는 `comments`의 key가 되지 않으니 태현님 말처럼 `@RequestParam` 을 사용하는 것이 맞는 것 같아요!! 

```
어떤 resource를 식별하고 싶으면 Path Variable를, 정렬이나 필터링을 한다면 Query Parameter를 사용하는 것이 좋다.

PathVariable은 REST API에서 값을 호출할 때 주로 사용되고, QueryParameter는 게시판의 페이지 및 검색 정보를 전달하는 방식에서 많이 사용된다.
```

taehyunnn reviewed (commented):
옥상에서는 약한 소리 하시더니 실제로는 굉장히 잘 작성하셨네요!! ㅎ
댓글까지 요구사항 추가하시다니 멋집니다~~
덕분에 새로운거 많이 배워갑니다!! 부리부리 대마왕 화이팅~ 👍

buri-1029 committed:
refactor : @Transactional 관련 피드백 반영

buri-1029 committed:
refactor : Post 관련 엔드포인트 Restful 하도록 변경

userId 파라미터로 넘기지 않고 userId가 담긴 dto를 통해
유저정보를 전달하여 가입유저를 확인하는 방식으로 구현

buri-1029 committed:
refactor : User 관련 엔드포인트 변경, 변경감지를 통해 삭제되도록 변경

buri-1029 committed:
refactor : Comment 관련 엔드포인트 변경

comment 패키지 변경, 해당 댓글이 매핑된 게시글에 포함된 댓글인지 확인하는 검증 로직 추가

buri-1029 committed:
refactor : ResponseExceptionHandler 수정

buri-1029 committed:
refactor : 코드 리팩토링

  - controller, service 메서드명 변경
  - 오류메세지에 관련 식별자 포함하도록 변경
  - delete api 반환 값 HttpStatus.NO_CONTENT으로 변경
  - RequestDto의 @NotNull -> @NotBlank으로 변경

buri-1029 commented on springboot-board/src/main/java/org/prgms/board/comment/service/CommentService.java:
```diff
@@ -0,0 +1,82 @@
+
+        if (!user.equals(comment.getWriter())) {
+            throw new NotMatchException("자신의 댓글만 삭제할 수 있습니다.");
+        }
```
통일된 메세지를 사용해서 메서드로 따로 빼니 코드가 깔꼬롬해졌어요👍 피드백 감사합니당 :)

buri-1029 committed:
test : 테스크코드 리택토링

buri-1029 committed:
refactor : dto validation 변경

buri-1029 committed:
update README.md

sunH0 commented on springboot-board/src/main/java/org/prgms/board/post/controller/PostController.java:
```diff
@@ -0,0 +1,82 @@
+
+    @PostMapping("/{id}/comments")
+    public ApiResponse<Long> writeComment(@PathVariable Long id,
+                                          @RequestBody @Valid CommentRequest commentRequest) {
```
@Valid 를 사용하면 NotNull 이나 NotBlank를 검증하주는 것이나요??

buri-1029 replied:
넹! Validation를 제공하는 어노테이션만 선언했다고 해서 알아서 검증해주지 않고 검증할 파라미터 앞에 `@Valid` 를 추가해야 검증을 할 수 있습니당 :)

sunH0 reviewed (commented):
부희님 과제하느라 수고하셨습니다~ 코드보고 저도 많이 배우게 됬어요
궁금한점이 있어서 답변해주시면 감사하겠습니다!

buri-1029 committed:
MySQL password 정보 환경변수 파일로 접근 변경

buri-1029 committed:
refactor : getPostsByUser API 엔드포인트 변경

0923kdh commented on springboot-board/src/main/java/org/prgms/board/common/response/ErrorResponse.java:
```diff
@@ -0,0 +1,38 @@
+
+@Getter
+public class ErrorResponse {
+    private int status;
+    private String error;
+    private String message;
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss", timezone = "Asia/Seoul")
+    private LocalDateTime serverDatetime;
```
필드 사이에 개행이 들어가면 좋을 것 같습니다~

0923kdh commented on springboot-board/src/main/java/org/prgms/board/common/response/ErrorResponse.java:
```diff
@@ -0,0 +1,38 @@
+    private int status;
+    private String error;
+    private String message;
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss", timezone = "Asia/Seoul")
```
timezone = "Asia/Seoul" 으로 설정한 이유가 있을까요 ? 기본값은 utc로 알고 있습니다!

0923kdh commented on springboot-board/src/main/java/org/prgms/board/common/response/ResponseExceptionHandler.java:
```diff
@@ -0,0 +1,29 @@
+    @ExceptionHandler({MethodArgumentNotValidException.class})
+    public ResponseEntity<ErrorResponse> handleMethodArgumentNotValid(MethodArgumentNotValidException ex) {
+        return ErrorResponse.toResponseEntity(HttpStatus.BAD_REQUEST,
+            ex.getBindingResult().getAllErrors().get(0).getDefaultMessage());
```
필드 정보도 같이 내려주면 좋을 것 같습니다!

0923kdh commented on springboot-board/src/main/java/org/prgms/board/domain/entity/Comment.java:
```diff
@@ -0,0 +1,45 @@
+    private Long id;
+
+    @Lob
+    @Column(name = "content", nullable = false)
```
@Column(name = "content", nullable = false, columnDefinition = "TEXT") private String content;
로 해주어야 nullable = false가 되는걸로 알고 있습니다!

0923kdh commented on springboot-board/src/main/java/org/prgms/board/domain/entity/Comment.java:
```diff
@@ -0,0 +1,45 @@
+        this.writer = writer;
+    }
+
+    public void changeInfo(String content) {
```
저도 처음에 changeInfo라는 메서드명을 사용했는데요. info라는 단어가 광범위(?)해서 메서드명 자체가 조금 모호하다는 리뷰를 받았었습니다! 부희님 생각은 어떠신가요?

0923kdh commented on springboot-board/src/main/java/org/prgms/board/post/controller/PostController.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    @DeleteMapping("/{id}")
+    public ApiResponse<Void> removePost(@PathVariable Long id,
```
제가 잘 몰라서 그러는데.. ApiResponse<Void> 랑 ApiResponse<void>  차이가 있나요 ..?

0923kdh commented on springboot-board/src/main/java/org/prgms/board/user/dto/UserRequest.java:
```diff
@@ -0,0 +1,34 @@
+
+@Getter
+public class UserRequest {
+    @NotBlank(message = "이름을 입력해주세요")
```
이렇게 되면 숫자나, 특수문자 와 같은 경우는 검증이 안되기때문에 custom annotation을 만들어서 검증해도 좋을 것 같습니다!

0923kdh commented on springboot-board/src/main/resources/application.yml:
```diff
@@ -0,0 +1,13 @@
+spring:
+  datasource:
+    url: jdbc:mysql://localhost:3307/dev_board?serverTimezone=Asia/Seoul&useUniCode=yes&characterEncoding=UTF-8
```
저같은 경우에는 url, username 모두 .env 으로 관리해주었습니다..!!

buri-1029 committed:
Update README.md

buri-1029 committed:
Create github-action.yml

buri-1029 committed:
Update .gitignore

buri-1029 committed:
Update .github 위치 이동

buri-1029 committed:
Update github-action.yml

buri-1029 committed:
Update github-action.yml

buri-1029 committed:
Update .github 원위치로 이동

buri-1029 committed:
Update github-action.yml

buri-1029 committed:
Delete github-action.yml

buri-1029 committed:
refactor : dirty checking 으로 댓글 등록
- swagger 설정
